packages/experimental-utils/src/ts-eslint/Rule.ts

```diff
@@ -248,7 +248,9 @@ interface RuleContext<
 
 // This isn't the correct signature, but it makes it easier to do custom unions within reusable listeners
 // never will break someone's code unless they specifically type the function argument
-type RuleFunction<T extends TSESTree.BaseNode = never> = (node: T) => void;
+type RuleFunction<T extends TSESTree.Node | TSESTree.Token = never> = (
```

Comment:
I missed this before - it's not possible for you to create a selector on a token - so we don't need the `TSESTree.Token` type here.

Also it's been a while - but I think that when I defined this I didn't use the `TSESTree.Node` type for two reasons: (1) because it means if you've got some custom AST nodes from another parser - you can't use them, and (2) IIRC there was some perf issues with TS due to the massive size of the `Node` union.

Reply:
I'll export `TokenBase` and do `TSESTree.NodeBase | TSESTree.TokenBase` instead

Reply:
Or should I do it with `NodeAndTokenData` instead?

Reply:
we don't need the token type at all because tokens aren't valid - so you can keep it as just `BaseNode`

Reply:
We need `Token` as we're also having a `RuleListener` for `Comment` & `Token`

https://github.com/typescript-eslint/typescript-eslint/blob/abfc19bf9364d881bdf594ee166a1deb23240630/packages/experimental-utils/src/ts-eslint/Rule.ts#L280
https://github.com/typescript-eslint/typescript-eslint/blob/abfc19bf9364d881bdf594ee166a1deb23240630/packages/experimental-utils/src/ts-eslint/Rule.ts#L341

Reply:
Weird. I don't think either of those actually work.
I think they were added at some point - but yeah a quick test doesn't look like they work.

Reply:
So I should delete these 2 then?

Reply:
Commented them out (for now) and nothing seems to break

Reply:
Should I just remove them?
Or what do you think is the best approach now?

Reply:
@bradzacher Should I just remove them?
Or what do you think is the best approach now?